File: repository/investment.go
package repository

import (
	"context"
	"database/sql"
	"fmt"
	"time"

	"github.com/google/uuid"
	"github.com/mohan9182/financer/api"
)

type InvestmentRepository struct {
	*BaseRepository
}

func NewInvestmentRepository(base *BaseRepository) *InvestmentRepository {
	return &InvestmentRepository{BaseRepository: base}
}

const investmentCols = `id, user_id, symbol, name, investment_type, current_price, prev_close, last_quote_at, manual_nav, created_at`

func scanInvestment(row scannable) (*api.InvestmentResponse, error) {
	var inst api.InvestmentResponse
	var uid, sym, lq sql.NullString
	var cur, prev, nav sql.NullFloat64
	var itype int
	var createdAt time.Time
	err := row.Scan(&inst.Id, &uid, &sym, &inst.Name, &itype, &cur, &prev, &lq, &nav, &createdAt)
	if err != nil {
		return nil, err
	}
	inst.Symbol = sym.String
	inst.InvestmentType = api.InvestmentType(itype)
	inst.CurrentPrice = float32(cur.Float64)
	inst.PrevClose = float32(prev.Float64)
	inst.ManualNav = float32(nav.Float64)
	if lq.Valid {
		if t, err := time.Parse(time.RFC3339, lq.String); err == nil {
			inst.LastQuoteAt = t
		}
	}
	inst.CreatedAt = createdAt
	return &inst, nil
}

func (r *InvestmentRepository) CreateInvestment(ctx context.Context, userID string, symbol, name string, it api.InvestmentType, manualNav float32) (*api.InvestmentResponse, error) {
	id := uuid.New().String()
	_, err := r.ExecContext(ctx,
		`INSERT INTO investments (id, user_id, symbol, name, investment_type, current_price, prev_close, last_quote_at, manual_nav, created_at)
		 VALUES (?, ?, ?, ?, ?, NULL, NULL, NULL, ?, ?)`,
		id, userID, sql.NullString{String: symbol, Valid: symbol != ""}, name, int(it),
		sql.NullFloat64{Float64: float64(manualNav), Valid: manualNav != 0},
		time.Now().UTC(),
	)
	if err != nil {
		return nil, fmt.Errorf("inserting investment: %w", err)
	}
	return r.GetInvestment(ctx, id)
}

func (r *InvestmentRepository) GetInvestment(ctx context.Context, id string) (*api.InvestmentResponse, error) {
	return QueryOne(ctx, r.readDB,
		`SELECT `+investmentCols+` FROM investments WHERE id = ?`,
		scanInvestment, id,
	)
}

func (r *InvestmentRepository) ListInvestments(ctx context.Context, userID string) ([]*api.InvestmentResponse, error) {
	return QueryAll(ctx, r.readDB,
		`SELECT `+investmentCols+` FROM investments WHERE user_id = ? ORDER BY created_at DESC`,
		scanInvestment, userID,
	)
}

func (r *InvestmentRepository) UpdateInvestment(ctx context.Context, id string, symbol, name string, it api.InvestmentType, manualNav float32) (*api.InvestmentResponse, error) {
	var oldSym sql.NullString
	r.readDB.QueryRowContext(ctx, `SELECT symbol FROM investments WHERE id = ?`, id).Scan(&oldSym)

	var sym sql.NullString
	if symbol != "" {
		sym = sql.NullString{String: symbol, Valid: true}
	}
	_, err := r.ExecContext(ctx,
		`UPDATE investments SET symbol = COALESCE(?, symbol), name = ?, investment_type = ?, manual_nav = ? WHERE id = ?`,
		sym, name, int(it),
		sql.NullFloat64{Float64: float64(manualNav), Valid: manualNav != 0},
		id,
	)
	if err != nil {
		return nil, fmt.Errorf("updating investment: %w", err)
	}
	// The cached price history belongs to the old symbol; drop it when the
	// symbol changed so it refetches under the new ticker.
	if symbol != "" && oldSym.String != symbol {
		if _, err := r.ExecContext(ctx, `DELETE FROM investment_price_history WHERE investment_id = ?`, id); err != nil {
			return nil, fmt.Errorf("clearing price history: %w", err)
		}
	}
	return r.GetInvestment(ctx, id)
}

func (r *InvestmentRepository) DeleteInvestment(ctx context.Context, id string) (bool, error) {
	if _, err := r.ExecContext(ctx, `DELETE FROM investment_price_history WHERE investment_id = ?`, id); err != nil {
		return false, fmt.Errorf("deleting investment: %w", err)
	}
	result, err := r.ExecContext(ctx, `DELETE FROM investments WHERE id = ?`, id)
	if err != nil {
		return false, fmt.Errorf("deleting investment: %w", err)
	}
	rows, _ := result.RowsAffected()
	return rows > 0, nil
}

func scanLot(row scannable) (*api.LotResponse, error) {
	var lot api.LotResponse
	var side int
	var occurredAt, createdAt time.Time
	err := row.Scan(&lot.Id, &lot.InvestmentId, &side, &lot.Quantity, &lot.Price, &occurredAt, &createdAt)
	if err != nil {
		return nil, err
	}
	lot.Side = int32(side)
	lot.OccurredAt = occurredAt
	lot.CreatedAt = createdAt
	return &lot, nil
}

func (r *InvestmentRepository) ListLots(ctx context.Context, investmentID string) ([]*api.LotResponse, error) {
	return QueryAll(ctx, r.readDB,
		`SELECT id, investment_id, side, quantity, price, occurred_at, created_at FROM investment_lots WHERE investment_id = ? ORDER BY occurred_at ASC, created_at ASC`,
		scanLot, investmentID,
	)
}

func (r *InvestmentRepository) CreateLot(ctx context.Context, userID, investmentID string, side int32, quantity, price float32, occurredAt time.Time) (*api.LotResponse, error) {
	id := uuid.New().String()
	now := time.Now().UTC()
	_, err := r.ExecContext(ctx,
		`INSERT INTO investment_lots (id, user_id, investment_id, side, quantity, price, occurred_at, created_at) VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
		id, userID, investmentID, side, quantity, price, occurredAt, now,
	)
	if err != nil {
		return nil, fmt.Errorf("inserting lot: %w", err)
	}
	return &api.LotResponse{
		Id:           id,
		InvestmentId: investmentID,
		Side:         side,
		Quantity:     quantity,
		Price:        price,
		OccurredAt:   occurredAt,
		CreatedAt:    now,
	}, nil
}

func (r *InvestmentRepository) DeleteLot(ctx context.Context, id string) (bool, error) {
	result, err := r.ExecContext(ctx, `DELETE FROM investment_lots WHERE id = ?`, id)
	if err != nil {
		return false, fmt.Errorf("deleting lot: %w", err)
	}
	rows, _ := result.RowsAffected()
	return rows > 0, nil
}

func (r *InvestmentRepository) UpdateQuote(ctx context.Context, id string, currentPrice, prevClose float32) error {
	_, err := r.ExecContext(ctx,
		`UPDATE investments SET current_price = ?, prev_close = ?, last_quote_at = ? WHERE id = ?`,
		currentPrice, prevClose, time.Now().UTC(), id,
	)
	return err
}

func (r *InvestmentRepository) ListAllForRefresh(ctx context.Context) ([]*api.InvestmentResponse, error) {
	return QueryAll(ctx, r.readDB,
		`SELECT `+investmentCols+` FROM investments WHERE symbol IS NOT NULL AND symbol != ''`,
		scanInvestment,
	)
}

func (r *InvestmentRepository) UpsertPriceHistory(ctx context.Context, investmentID, rangeID string, ts []int64, closes []float64, fetchedAt int64) error {
	return r.ExecInTx(ctx, func(tx *Tx) error {
		if _, err := tx.ExecContext(ctx, `DELETE FROM investment_price_history WHERE investment_id = ? AND range_id = ?`, investmentID, rangeID); err != nil {
			return err
		}
		stmt, err := tx.PrepareContext(ctx, `INSERT INTO investment_price_history (investment_id, range_id, t, close, fetched_at) VALUES (?, ?, ?, ?, ?)`)
		if err != nil {
			return err
		}
		defer stmt.Close()
		for i := range ts {
			if _, err := stmt.ExecContext(ctx, investmentID, rangeID, ts[i], closes[i], fetchedAt); err != nil {
				return err
			}
		}
		return nil
	})
}

func (r *InvestmentRepository) GetPriceHistory(ctx context.Context, investmentID, rangeID string) (ts []int64, closes []float64, lastFetched int64, err error) {
	rows, err := r.readDB.QueryContext(ctx,
		`SELECT t, close, fetched_at FROM investment_price_history WHERE investment_id = ? AND range_id = ? ORDER BY t`,
		investmentID, rangeID,
	)
	if err != nil {
		return nil, nil, 0, err
	}
	defer rows.Close()
	for rows.Next() {
		var t int64
		var c float64
		var fa int64
		if err := rows.Scan(&t, &c, &fa); err != nil {
			return nil, nil, 0, err
		}
		ts = append(ts, t)
		closes = append(closes, c)
		lastFetched = fa
	}
	return ts, closes, lastFetched, rows.Err()
}

File: src/repo/investment.rs
//! Investments repository — CRUD + lots + price history on `SurrealDB`,
//! mirroring the Go backend's `repository/investment.go`.

use surrealdb::Connection;
use utoipa::ToSchema;

use crate::error::Result;
use crate::repo::surreal::{rid, take_json, DbClient, RepoConn};
use crate::timex::{go_ts, round2, ts_rfc3339};

/// Investment type. Wire string (`STOCK`/`MUTUAL_FUND`), stored as-is.
#[derive(
    Debug, Clone, Copy, PartialEq, Eq, serde::Serialize, serde::Deserialize, strum::Display, strum::EnumString, ToSchema,
)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
#[strum(serialize_all = "SCREAMING_SNAKE_CASE")]
#[schema(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum InvestmentType {
    Stock,
    MutualFund,
}

#[derive(Clone)]
pub struct InvestmentRepo<C: Connection = DbClient> {
    db: RepoConn<C>,
}

impl<C: Connection> InvestmentRepo<C> {
    pub fn new(db: RepoConn<C>) -> Self {
        Self { db }
    }

    pub async fn create_investment(&self, user_id: &str, symbol: &str, name: &str, it: InvestmentType, manual_nav: f64) -> Result<InvestmentRow> {
        let now = go_ts(chrono::Utc::now());
        let sym = if symbol.is_empty() { None } else { Some(symbol.to_string()) };
        let nav = if manual_nav == 0.0 { None } else { Some(manual_nav) };
        let mut res = self
            .db
            .query(
                "CREATE investment CONTENT {
                    user: $uid, symbol: $sym, name: $name, investmentType: $it,
                    currentPrice: 0.0, prevClose: 0.0, lastQuoteAt: NONE,
                    manualNav: $nav, createdAt: $created
                } RETURN meta::id(id) AS id, symbol, name, investmentType,
                    currentPrice, prevClose, manualNav, lastQuoteAt, createdAt",
            )
            .bind(("uid", rid("user", user_id)))
            .bind(("sym", sym))
            .bind(("name", name.to_string()))
            .bind(("it", it.to_string()))
            .bind(("nav", nav))
            .bind(("created", now.clone()))
            .await?
            .check()?;
        let mut row = take_json::<InvestmentRow>(&mut res, 0)?
            .into_iter()
            .next()
            .ok_or_else(|| crate::error::ApiError::internal("investment create returned no row"))?;
        row.created_at = now;
        Ok(row)
    }

    pub async fn get_investment(&self, user_id: &str, id: &str) -> Result<Option<InvestmentRow>> {
        let mut res = self
            .db
            .query(
                "SELECT meta::id(id) AS id, symbol, name, investmentType,
                    currentPrice, prevClose, manualNav, lastQuoteAt, createdAt
                 FROM investment WHERE id = $rid AND user = $uid LIMIT 1",
            )
            .bind(("rid", rid("investment", id)))
            .bind(("uid", rid("user", user_id)))
            .await?;
        Ok(take_json(&mut res, 0)?.into_iter().next())
    }

    pub async fn get_by_symbol(&self, user_id: &str, symbol: &str) -> Result<Option<InvestmentRow>> {
        if symbol.is_empty() {
            return Ok(None);
        }
        let mut res = self
            .db
            .query(
                "SELECT meta::id(id) AS id, symbol, name, investmentType,
                    currentPrice, prevClose, manualNav, lastQuoteAt, createdAt
                 FROM investment WHERE user = $uid AND symbol = $sym LIMIT 1",
            )
            .bind(("uid", rid("user", user_id)))
            .bind(("sym", symbol.to_string()))
            .await?;
        Ok(take_json(&mut res, 0)?.into_iter().next())
    }

    pub async fn list_investments(&self, user_id: &str) -> Result<Vec<InvestmentRow>> {
        let mut res = self
            .db
            .query(
                "SELECT meta::id(id) AS id, symbol, name, investmentType,
                    currentPrice, prevClose, manualNav, lastQuoteAt, createdAt
                 FROM investment WHERE user = $uid ORDER BY createdAt DESC",
            )
            .bind(("uid", rid("user", user_id)))
            .await?;
        Ok(take_json(&mut res, 0)?)
    }

    pub async fn update_investment(&self, user_id: &str, id: &str, symbol: &str, name: &str, it: InvestmentType, manual_nav: f64) -> Result<InvestmentRow> {
        let old_sym: Option<String> = {
            let mut res = self
                .db
                .query("SELECT VALUE symbol FROM investment WHERE id = $rid AND user = $uid LIMIT 1")
                .bind(("rid", rid("investment", id)))
                .bind(("uid", rid("user", user_id)))
                .await?;
            take_json::<String>(&mut res, 0)?.into_iter().next()
        };
        let sym: Option<String> = if symbol.is_empty() { None } else { Some(symbol.to_string()) };
        let nav: Option<f64> = if manual_nav == 0.0 { None } else { Some(manual_nav) };
        self.db
            .query(
                "UPDATE $rid SET
                    symbol = IF $sym != NONE THEN $sym ELSE symbol END,
                    name = $name, investmentType = $it, manualNav = IF $nav != NONE THEN $nav ELSE manualNav END
                 WHERE user = $uid",
            )
            .bind(("rid", rid("investment", id)))
            .bind(("uid", rid("user", user_id)))
            .bind(("sym", sym))
            .bind(("name", name.to_string()))
            .bind(("it", it.to_string()))
            .bind(("nav", nav))
            .await?
            .check()?;
        // Drop cached price history when the symbol changed so it refetches.
        if !symbol.is_empty() && old_sym.as_deref() != Some(symbol) {
            self.db
                .query("DELETE investment_price_history WHERE investment = $rid")
                .bind(("rid", rid("investment", id)))
                .await?
                .check()?;
        }
        Ok(self.get_investment(user_id, id).await?.unwrap())
    }

    pub async fn delete_investment(&self, user_id: &str, id: &str) -> Result<bool> {
        self.db
            .query("DELETE investment_price_history WHERE investment = $rid")
            .bind(("rid", rid("investment", id)))
            .await?
            .check()?;
        let mut res = self
            .db
            .query("DELETE $rid WHERE user = $uid RETURN BEFORE")
            .bind(("rid", rid("investment", id)))
            .bind(("uid", rid("user", user_id)))
            .await?;
        Ok(!take_json::<serde_json::Value>(&mut res, 0)?.is_empty())
    }

    pub async fn list_lots(&self, user_id: &str, investment_id: &str) -> Result<Vec<LotRow>> {
        let mut res = self
            .db
            .query(
                "SELECT meta::id(id) AS id, meta::id(investment) AS investmentId, side,
                    quantity, price, occurredAt, createdAt
                 FROM investment_lot
                 WHERE investment = $rid AND user = $uid
                 ORDER BY occurredAt ASC, createdAt ASC",
            )
            .bind(("rid", rid("investment", investment_id)))
            .bind(("uid", rid("user", user_id)))
            .await?;
        Ok(take_json(&mut res, 0)?)
    }

    pub async fn create_lot(&self, user_id: &str, investment_id: &str, side: i64, quantity: f64, price: f64, occurred_at: &str) -> Result<LotRow> {
        let now = go_ts(chrono::Utc::now());
        let mut res = self
            .db
            .query(
                "CREATE investment_lot CONTENT {
                    user: $uid, investment: $rid, side: $side, quantity: $quantity,
                    price: $price, occurredAt: $occ, createdAt: $created
                } RETURN meta::id(id) AS id, meta::id(investment) AS investmentId,
                    side, quantity, price, occurredAt, createdAt",
            )
            .bind(("uid", rid("user", user_id)))
            .bind(("rid", rid("investment", investment_id)))
            .bind(("side", side))
            .bind(("quantity", quantity))
            .bind(("price", price))
            .bind(("occ", occurred_at.to_string()))
            .bind(("created", now.clone()))
            .await?
            .check()?;
        let mut lot = take_json::<LotRow>(&mut res, 0)?
            .into_iter()
            .next()
            .ok_or_else(|| crate::error::ApiError::internal("lot create returned no row"))?;
        lot.created_at = now;
        Ok(lot)
    }

    pub async fn delete_lot(&self, user_id: &str, id: &str) -> Result<bool> {
        let mut res = self
            .db
            .query("DELETE $rid WHERE user = $uid RETURN BEFORE")
            .bind(("rid", rid("investment_lot", id)))
            .bind(("uid", rid("user", user_id)))
            .await?;
        Ok(!take_json::<serde_json::Value>(&mut res, 0)?.is_empty())
    }

    /// Idempotent insert: a second insert with the same (user, `externalId`) is
    /// skipped. Returns whether the row was inserted.
    pub async fn insert_lot(&self, user_id: &str, investment_id: &str, input: &LotInput) -> Result<bool> {
        let now = go_ts(chrono::Utc::now());
        let ext: Option<String> = if input.external_id.is_empty() { None } else { Some(input.external_id.clone()) };
        // Dedup check (matches the old INSERT OR IGNORE on (user, externalId)).
        if let Some(e) = &ext {
            let mut res = self
                .db
                .query("SELECT VALUE meta::id(id) FROM investment_lot WHERE user = $uid AND externalId = $ext LIMIT 1")
                .bind(("uid", rid("user", user_id)))
                .bind(("ext", e.clone()))
                .await?;
            if !take_json::<String>(&mut res, 0)?.is_empty() {
                return Ok(false);
            }
        }
        self.db
            .query(
                "CREATE investment_lot CONTENT {
                    user: $uid, investment: $rid, side: $side, quantity: $quantity,
                    price: $price, occurredAt: $occ, createdAt: $created, externalId: $ext
                }",
            )
            .bind(("uid", rid("user", user_id)))
            .bind(("rid", rid("investment", investment_id)))
            .bind(("side", input.side))
            .bind(("quantity", input.quantity))
            .bind(("price", input.price))
            .bind(("occ", input.occurred_at.clone()))
            .bind(("created", now))
            .bind(("ext", ext))
            .await?
            .check()?;
        Ok(true)
    }

    pub async fn update_lot(&self, user_id: &str, id: &str, quantity: f64, price: f64, occurred_at: &str) -> Result<bool> {
        let mut res = self
            .db
            .query(
                "UPDATE $rid SET quantity = $q, price = $p, occurredAt = $occ
                 WHERE user = $uid RETURN meta::id(id) AS id",
            )
            .bind(("rid", rid("investment_lot", id)))
            .bind(("uid", rid("user", user_id)))
            .bind(("q", quantity))
            .bind(("p", price))
            .bind(("occ", occurred_at.to_string()))
            .await?;
        Ok(!take_json::<serde_json::Value>(&mut res, 0)?.is_empty())
    }

    pub async fn update_quote(&self, id: &str, current_price: f64, prev_close: f64) -> Result<()> {
        self.db
            .query(
                "UPDATE $rid SET currentPrice = $price, prevClose = $prev, lastQuoteAt = $at",
            )
            .bind(("rid", rid("investment", id)))
            .bind(("price", current_price))
            .bind(("prev", prev_close))
            .bind(("at", go_ts(chrono::Utc::now())))
            .await?
            .check()?;
        Ok(())
    }

    /// Replace the cached price history for (investment, range).
    pub async fn upsert_price_history(&self, investment_id: &str, range_id: &str, ts: &[i64], closes: &[f64], fetched_at: i64) -> Result<()> {
        let rid_inv = rid("investment", investment_id);
        self.db
            .query("DELETE investment_price_history WHERE investment = $rid AND rangeId = $range")
            .bind(("rid", rid_inv.clone()))
            .bind(("range", range_id.to_string()))
            .await?
            .check()?;
        for (i, t) in ts.iter().enumerate() {
            self.db
                .query(
                    "CREATE investment_price_history CONTENT {
                        investment: $rid, rangeId: $range, t: $t, close: $close, fetchedAt: $fetched
                    }",
                )
                .bind(("rid", rid_inv.clone()))
                .bind(("range", range_id.to_string()))
                .bind(("t", *t))
                .bind(("close", closes.get(i).copied().unwrap_or(0.0)))
                .bind(("fetched", fetched_at))
                .await?
                .check()?;
        }
        Ok(())
    }

    /// Read cached price history: ts, closes, last fetched timestamp.
    pub async fn get_price_history(&self, user_id: &str, investment_id: &str, range_id: &str) -> Result<(Vec<i64>, Vec<f64>, i64)> {
        let mut res = self
            .db
            .query(
                "SELECT t, close, fetchedAt FROM investment_price_history
                 WHERE investment = $rid AND rangeId = $range
                   AND investment IN (SELECT VALUE id FROM investment WHERE user = $uid)
                 ORDER BY t",
            )
            .bind(("rid", rid("investment", investment_id)))
            .bind(("range", range_id.to_string()))
            .bind(("uid", rid("user", user_id)))
            .await?;
        let rows: Vec<PriceHistoryRow> = take_json(&mut res, 0)?;
        let mut ts = Vec::new();
        let mut closes = Vec::new();
        let mut fetched = 0i64;
        for r in rows {
            ts.push(r.t);
            closes.push(r.close);
            fetched = r.fetched_at;
        }
        Ok((ts, closes, fetched))
    }
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct InvestmentRow {
    pub id: String,
    #[serde(default)]
    pub symbol: String,
    pub name: String,
    #[allow(clippy::struct_field_names)]
    pub investment_type: InvestmentType,
    #[serde(default, deserialize_with = "de_null_f64")]
    pub current_price: f64,
    #[serde(default, deserialize_with = "de_null_f64")]
    pub prev_close: f64,
    #[serde(default, deserialize_with = "de_null_f64")]
    pub manual_nav: f64,
    #[serde(default, deserialize_with = "de_null_string")]
    pub last_quote_at: String,
    pub created_at: String,
}

fn de_null_string<'de, D: serde::Deserializer<'de>>(d: D) -> std::result::Result<String, D::Error> {
    let v: Option<String> = serde::Deserialize::deserialize(d)?;
    Ok(v.unwrap_or_default())
}

fn de_null_f64<'de, D: serde::Deserializer<'de>>(d: D) -> std::result::Result<f64, D::Error> {
    let v: Option<f64> = serde::Deserialize::deserialize(d)?;
    Ok(v.unwrap_or(0.0))
}

#[derive(serde::Deserialize)]
struct PriceHistoryRow {
    t: i64,
    close: f64,
    #[serde(rename = "fetchedAt")]
    fetched_at: i64,
}

pub struct LotInput {
    pub side: i64,
    pub quantity: f64,
    pub price: f64,
    pub occurred_at: String,
    pub external_id: String,
}

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LotRow {
    pub id: String,
    pub investment_id: String,
    pub side: i64, // 1 buy, -1 sell
    pub quantity: f64,
    pub price: f64,
    pub occurred_at: String,
    pub created_at: String,
}

/// Wire investment (serde covers Go's `investmentWire`).
#[derive(Debug, Clone, serde::Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct Investment {
    pub id: String,
    pub symbol: String,
    pub name: String,
    #[allow(clippy::struct_field_names)]
    pub investment_type: InvestmentType,
    pub current_price: f64,
    pub prev_close: f64,
    pub manual_nav: f64,
    #[serde(rename = "lastQuoteAt")]
    pub last_quote_at: String,
    #[serde(rename = "createdAt")]
    pub created_at: String,
    pub quantity: f64,
    pub avg_cost: f64,
    pub current_value: f64,
    pub unrealized_pnl: f64,
    pub realized_pnl: f64,
}

/// Wire lot.
#[derive(Debug, Clone, serde::Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct Lot {
    pub id: String,
    #[serde(rename = "investmentId")]
    pub investment_id: String,
    pub side: i64,
    pub quantity: f64,
    pub price: f64,
    #[serde(rename = "occurredAt")]
    pub occurred_at: String,
    #[serde(rename = "createdAt")]
    pub created_at: String,
}

/// Wire investment with position filled in (Go's `investmentWire` + `withPosition`).
pub fn investment_wire(r: &InvestmentRow, qty: f64, avg_cost: f64, current_value: f64, unrealized: f64, realized: f64) -> Investment {
    let price = effective_price(r);
    Investment {
        id: r.id.clone(),
        symbol: r.symbol.clone(),
        name: r.name.clone(),
        investment_type: r.investment_type,
        current_price: round2(price),
        prev_close: round2(r.prev_close),
        manual_nav: round2(r.manual_nav),
        last_quote_at: if r.last_quote_at.is_empty() { String::new() } else { ts_rfc3339(&r.last_quote_at) },
        created_at: ts_rfc3339(&r.created_at),
        quantity: round2(qty),
        avg_cost: round2(avg_cost),
        current_value: round2(current_value),
        unrealized_pnl: round2(unrealized),
        realized_pnl: round2(realized),
    }
}

pub fn lot_wire(l: &LotRow) -> Lot {
    Lot {
        id: l.id.clone(),
        investment_id: l.investment_id.clone(),
        side: l.side,
        quantity: round2(l.quantity),
        price: round2(l.price),
        occurred_at: ts_rfc3339(&l.occurred_at),
        created_at: ts_rfc3339(&l.created_at),
    }
}

/// effective price: `manual_nav` if set, else cached Yahoo price.
pub fn effective_price(r: &InvestmentRow) -> f64 {
    if r.manual_nav > 0.0 {
        r.manual_nav
    } else {
        r.current_price
    }
}

#[cfg(test)]
mod tests {
    use std::sync::Arc;
    use surrealdb::Surreal;

    use super::*;
    use crate::surreal_db;

    async fn repo() -> (InvestmentRepo<surrealdb::engine::local::Db>, Arc<surrealdb::Surreal<surrealdb::engine::local::Db>>) {
        let db = Arc::new(surreal_db::connect_mem().await.unwrap());
        db.query("CREATE user CONTENT { id: 'u1', email: 'u1@x.com', passwordHash: 'h', name: 'u1' }")
            .await
            .unwrap()
            .check()
            .unwrap();
        (InvestmentRepo::new(db.clone()), db)
    }



    #[tokio::test]
    async fn investment_crud_and_lots() {
        let (repo, _db) = repo().await;
        let inst = repo.create_investment("u1", "RELIANCE", "Reliance", InvestmentType::Stock, 0.0).await.unwrap();
        assert_eq!(inst.symbol, "RELIANCE");
        assert_eq!(inst.investment_type, InvestmentType::Stock);

        let dup = repo.get_by_symbol("u1", "RELIANCE").await.unwrap();
        assert_eq!(dup.unwrap().id, inst.id);

        let lot = repo.create_lot("u1", &inst.id, 1, 10.0, 100.0, "2024-01-02 10:00:00 +0000 UTC").await.unwrap();
        assert_eq!(lot.side, 1);
        assert_eq!(lot.investment_id, inst.id);
        let lots = repo.list_lots("u1", &inst.id).await.unwrap();
        assert_eq!(lots.len(), 1);

        let updated = repo.update_lot("u1", &lot.id, 20.0, 110.0, "2024-01-02 10:00:00 +0000 UTC").await.unwrap();
        assert!(updated);
        assert_eq!(repo.list_lots("u1", &inst.id).await.unwrap()[0].quantity, 20.0);

        assert!(repo.delete_lot("u1", &lot.id).await.unwrap());
        assert!(repo.delete_investment("u1", &inst.id).await.unwrap());
        assert!(repo.get_investment("u1", &inst.id).await.unwrap().is_none());
    }
}
